### PR TITLE
WIP: SCAT-3090 - get messages from suppliers

### PR DESCRIPTION
**Pull request by sudhakar-reddy-konda**
### Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No

**Before creating a pull request make sure that:**

- [ ] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [ ] tests have been updated / new tests has been added (if needed)

### CODE REVIEW

File: src/main/java/uk/gov/crowncommercial/dts/scale/cat/controller/EventsController.java [hunk 1]
```diff
@@ -171,4 +171,20 @@ public StringValueResponse publishEvent(@PathVariable("procID") final Integer pr
     return new StringValueResponse("OK");
   }
 
+  @GetMapping("/{eventID}/messages")
+  public Collection<MessageSummary> getMessages(
+          @PathVariable("procID") final Integer procId,
+          @PathVariable("eventID") final String eventId,
+          @RequestParam(name = "message-direction",required = false) String messageDirection,
+          @RequestParam(name = "message-read",required = false) String messageRead,
+          @RequestParam(name = "sort") String sort,
```
Comment (tommyb82) [inline src/main/java/uk/gov/crowncommercial/dts/scale/cat/controller/EventsController.java:180]: These can be deserialized directly to the enum types I believe (as the generated types have `@JsonCreator` on their fromValue method - e.g. `MessageSort` - so should just work

File: src/main/java/uk/gov/crowncommercial/dts/scale/cat/service/JaggaerService.java [hunk 2]
```diff
@@ -216,15 +216,33 @@ public void publishRfx(final ProcurementEvent event, final PublishDates publishD
...
+
+  public MessagesResponse getMessages(final String externalEventId, final String messageDirection,
+      final String messageRead, final String sort, final Integer page,
+      final Integer pageSize) {
+    final var messagesUrl = jaggaerAPIConfig.getGetMessages().get(ENDPOINT);
+    final int start = page +(page > 1 ?pageSize:0);
+    final String filters = "objectReferenceCode=="+externalEventId;
+    final String queryPrams = "start="+start;
+
+
```
Comment (tommyb82) [inline src/main/java/uk/gov/crowncommercial/dts/scale/cat/service/JaggaerService.java:239]: Remove unnecessary blank lines

File: src/main/java/uk/gov/crowncommercial/dts/scale/cat/service/JaggaerService.java [hunk 3]
```diff
@@ -216,15 +216,33 @@ public void publishRfx(final ProcurementEvent event, final PublishDates publishD
...
-            .orElseThrow(() -> new JaggaerApplicationException(INTERNAL_SERVER_ERROR.value(),
-                    "Unexpected error retrieving projects"));
+        .retrieve()
+        .bodyToMono(ProjectListResponse.class)
+        .block(ofSeconds(jaggaerAPIConfig.getTimeoutDuration())))
+        .orElseThrow(() -> new JaggaerApplicationException(INTERNAL_SERVER_ERROR.value(),
+            "Unexpected error retrieving projects"));
+  }
+
+  public MessagesResponse getMessages(final String externalEventId, final String messageDirection,
```
Comment (tommyb82) [inline src/main/java/uk/gov/crowncommercial/dts/scale/cat/service/JaggaerService.java:231]: Assume we'll be making use of the unused params in follow up enhancements?

File: src/main/java/uk/gov/crowncommercial/dts/scale/cat/service/ProcurementEventService.java [hunk 4]
```diff
@@ -598,4 +602,79 @@ public void publishEvent(final Integer procId, final String eventId,
...
+   * @param procId
+   * @param eventId
+   * @param messageDirection
+   * @param messageRead
+   * @param sort
+   * @param page
+   * @param pageSize
+   * @return
+   */
+  public Collection<MessageSummary> getMessageSummaries(
```
Comment (tommyb82) [inline src/main/java/uk/gov/crowncommercial/dts/scale/cat/service/ProcurementEventService.java:617]: As with the controller - probably be good to start a new `MessagingService` or similar - perhaps liaise with @venkibric as he might have already done so 👍

File: src/main/java/uk/gov/crowncommercial/dts/scale/cat/service/ProcurementEventService.java [hunk 5]
```diff
@@ -598,4 +602,79 @@ public void publishEvent(final Integer procId, final String eventId,
...
+            messageDirection, messageRead, sort, page, pageSize);
+
+    return messagesResponse.getMessageList().getMessage()
+            .stream().map(message -> convertMessageToMessageSummary(message,pageSize,messagesResponse.getTotRecords()))
+            .collect(Collectors.toList());
+
+  }
+
+  private MessageSummary convertMessageToMessageSummary(final Message message, final Integer pageSize, final Integer totalRecords) {
+    var messageSummary = new MessageSummary();
```
Comment (tommyb82) [inline src/main/java/uk/gov/crowncommercial/dts/scale/cat/service/ProcurementEventService.java:632]: The generated types have chained setters - can reduce a bit of boilerplate e.g.
```
return new MessageSummary().OCDS(getCaTMessageOCDS(message))
        .nonOCDS(getMessageSummaryAllOfNonOCDS(message, pageSize, totalRecords));
```

**Review by tommyb82 — COMMENTED** [review]
Looking good - mostly minor comments (be good to ringfence the messaging stuff as much as possible as the events controller / service class is already quite congested) 👍

File: src/main/java/uk/gov/crowncommercial/dts/scale/cat/service/ProcurementEventService.java (hunk 4 above)
Comment (adrianmilne) [reply src/main/java/uk/gov/crowncommercial/dts/scale/cat/service/ProcurementEventService.java:617]: Yep - agree this would be good 👍

Comment (sudhakar-reddy-konda) [reply src/main/java/uk/gov/crowncommercial/dts/scale/cat/service/ProcurementEventService.java]: Venki created those, need to wait for his PR into develop

File: src/main/java/uk/gov/crowncommercial/dts/scale/cat/controller/EventsController.java [hunk 6]
```diff
@@ -171,4 +171,20 @@ public StringValueResponse publishEvent(@PathVariable("procID") final Integer pr
     return new StringValueResponse("OK");
   }
 
+  @GetMapping("/{eventID}/messages")
+  public Collection<MessageSummary> getMessages(
```
[the remark this answers is not in the document]
Comment (sudhakar-reddy-konda) [reply src/main/java/uk/gov/crowncommercial/dts/scale/cat/controller/EventsController.java]: Venki already done this, need wait for his changes merged into dev

File: src/main/java/uk/gov/crowncommercial/dts/scale/cat/controller/EventsController.java (hunk 1 above)
Comment (sudhakar-reddy-konda) [reply src/main/java/uk/gov/crowncommercial/dts/scale/cat/controller/EventsController.java]: done

File: src/main/java/uk/gov/crowncommercial/dts/scale/cat/service/JaggaerService.java (hunk 2 above)
Comment (sudhakar-reddy-konda) [reply src/main/java/uk/gov/crowncommercial/dts/scale/cat/service/JaggaerService.java]: done

**Review by tommyb82 — COMMENTED** [review]
Still reviewing - getting a 500 for a newly created project / event that has no messages - presumably should return an empty `MessageSummary` object?

**Review by tommyb82 — APPROVED** [review]
👍

**Pull request merged**